lock/common.go

```diff
@@ -0,0 +1,47 @@
+// Package lock provides implementations for cancellable and distributed locks.
+package lock
+
+import (
+	"time"
+
+	"gopkg.in/src-d/go-errors.v0"
+)
+
+var (
+	ErrCanceled      = errors.NewKind("context canceled")
+	ErrAlreadyClosed = errors.NewKind("already closed")
+)
+
+type SessionConfig struct {
+	TTL time.Duration
+}
+
+// Service is a locking service.
+type Service interface {
+	NewSession(*SessionConfig) (Session, error)
+	// Close closes the service and releases any resources held by it.
+	// If it is called more than once, ErrAlreadyClosed is returned.
+	Close() error
+}
+
+// Session is a locking session that can be reused to get multiple locks.
+// Multiple actors should use different sessions.
+type Session interface {
+	// NewService creates a NewService for the given id. Lockers returned for the
+	// same id on different sessions are mutually exclusive.
+	NewLocker(id string) Locker
```

Comment:
Nope. `NewLocker` is meant to be lazy if it can fail. Errors should be returned when getting a new session or when acquiring a lock. `NewLocker` just creates a handle that can be used to lock and an unlock, and those are the operations that could fail.